prov/verbs/src/verbs_ep.c

```diff
@@ -176,27 +176,26 @@ ssize_t vrb_post_send(struct vrb_ep *ep, struct ibv_send_wr *wr, uint64_t flags)
 	int ret, err;
 
 	ofi_genlock_lock(&vrb_ep2_progress(ep)->ep_lock);
-	ctx = vrb_alloc_ctx(vrb_ep2_progress(ep));
-	if (!ctx) {
-		ret = -FI_EAGAIN;
-		goto unlock;
-	}
 
-	if (!ep->sq_credits || !ep->peer_rq_credits) {
+	if (!ep->sq_credits) {
 		cq = container_of(ep->util_ep.tx_cq, struct vrb_cq, util_cq);
 		vrb_flush_cq(cq);
 
-		if (!ep->sq_credits || !ep->peer_rq_credits)
-			goto freectx;
+		if (!ep->sq_credits)
+			goto unlock;
 	}
 
-	if (vrb_wr_consumes_recv(wr) && !--ep->peer_rq_credits &&
-	    !(flags & OFI_PRIORITY)) {
-		/* Last credit is reserved for credit update */
-		ep->peer_rq_credits++;
-		goto freectx;
+	if (vrb_wr_consumes_recv(wr)) {
+		if ((ep->peer_rq_credits == 1) && !(flags & OFI_PRIORITY))
+			goto unlock;
+
+		ep->peer_rq_credits--;
```

Comment:
This flow isn't quite right.  peer_rq_credits could be 0 on entry into this call (because we used the last credit to send a flow control message).  The results of the above flow would result in value wrapping.  This check should work:
```
if (!ep->peer_rq_credits || ((ep->peer_rq_credits == 1) && !(flags & OFI_PRIORITY)))
    goto unlock;
```